rust/agama-server/src/storage/web/zfcp.rs

```diff
@@ -0,0 +1,254 @@
+//! This module implements the web API for the handling of zFCP storage service.
+//!
+//! The module offers two public functions:
+//!
+//! * `zfcp_service` which returns the Axum service.
+//! * `zfcp_stream` which offers an stream that emits the zFCP-related events coming from D-Bus.
+
+use agama_lib::{
+    error::ServiceError,
+    storage::{
+        client::zfcp::ZFCPClient,
+        model::zfcp::{ZFCPController, ZFCPDisk, ZFCPOptions},
+    },
+};
+use axum::{
+    extract::{Path, State},
+    routing::{get, post},
+    Json, Router,
+};
+use stream::{ZFCPControllerStream, ZFCPDiskStream};
+
+mod stream;
+
+use crate::{error::Error, web::common::EventStreams};
+
+/// Returns the stream of zFCP-related events.
+///
+/// The stream combines the following events:
+///
+/// * Changes on the zFCP devices collection.
+///
+/// * `dbus`: D-Bus connection to use.
+pub async fn zfcp_stream(dbus: &zbus::Connection) -> Result<EventStreams, Error> {
+    let stream: EventStreams = vec![
+        ("zfcp_disks", Box::pin(ZFCPDiskStream::new(dbus).await?)),
+        (
+            "zfcp_controllers",
+            Box::pin(ZFCPControllerStream::new(dbus).await?),
+        ),
+    ];
+    Ok(stream)
+}
+
+#[derive(Clone)]
+struct ZFCPState<'a> {
+    client: ZFCPClient<'a>,
+}
+
+pub async fn zfcp_service<T>(dbus: &zbus::Connection) -> Result<Router<T>, ServiceError> {
+    let client = ZFCPClient::new(dbus.clone()).await?;
+    let state = ZFCPState { client };
+    let router = Router::new()
+        .route("/supported", get(supported))
+        .route("/controllers", get(controllers))
+        .route(
+            "/controllers/:controller_id/activate",
+            post(activate_controller),
+        )
+        .route("/controllers/:controller_id/wwpns", get(get_wwpns))
+        .route(
+            "/controllers/:controller_id/wwpns/:wwpn_id/luns",
+            get(get_luns),
+        )
+        .route(
+            "/controllers/:controller_id/wwpns/:wwpn_id/luns/:lun_id/activate_disk",
```

Comment:
I am not sure about those URLs that only support POST actions. I wonder whether the arguments should be the regular payload of the POST method. Let's keep it as it is by now.